lib/dependencies.js

```diff
@@ -0,0 +1,384 @@
+          });
+        } else {
+          entries.push({
+            type: "stat",
+            relativePath: entry.relativePath,
+            size: entry.size,
+            mtime: +entry.mtime,
+            mode: entry.mode
+          });
+        }
+      }
+      fsTrees.push({
+        fsRoot,
+        entries
+      });
+    }
+    let serialized = {
+      rootDir: this.rootDir,
+      dependencies,
+      fsTrees
+    };
+    return serialized;
+  }
+
+  /**
+   * Deserialize from JSON data returned from the `serialize` method.
+   *
+   * @param dependencyData {ReturnType<Dependencies['serialize']>}
+   * @param [newRootDir] {string | undefined}
+   * @return {Dependencies};
+   */
+  static deserialize(dependencyData, newRootDir) {
+    let oldRootDir = dependencyData.rootDir;
+    newRootDir = newRootDir || oldRootDir;
+    let dependencies = new Dependencies(newRootDir);
+    let files = Object.keys(dependencyData.dependencies);
+    for (let file of files) {
+      let deps = dependencyData.dependencies[file];
+      if (newRootDir) {
+        for (let i = 0; i < deps.length; i++) {
+          let dep = deps[i];
+          if (dep.startsWith(oldRootDir+path.sep)) {
+            deps[i] = dep.replace(oldRootDir, newRootDir);
+          }
+        }
+      }
+      dependencies.setDependencies(file, deps);
+    }
+    /** @type {Map<string, FSTree>} */
+    let fsTrees = new Map();
+    for (let fsTreeData of dependencyData.fsTrees) {
+      /** @type {Array<Entry | HashEntry>} */
+      let entries = [];
+      for (let entry of fsTreeData.entries) {
+        if (entry.type === "stat") {
+          entries.push(new Entry(entry.relativePath, entry.size, entry.mtime, entry.mode));
+        } else {
+          entries.push(new HashEntry(entry.relativePath, entry.hash));
+        }
+      }
+      /** @type {FSTree | FSHashTree}  */
+      let fsTree;
+      /** @type {string} */
+      let treeRoot;
+      if (fsTreeData.fsRoot === oldRootDir) {
+        treeRoot = newRootDir;
+        fsTree = FSHashTree.fromHashEntries(entries, { sortAndExpand: true });
+      } else {
+        treeRoot = fsTreeData.fsRoot;
+        fsTree = FSTree.fromEntries(entries, { sortAndExpand: true });
+      }
+      fsTrees.set(treeRoot, fsTree);
+    }
+    dependencies.seal();
+    dependencies.fsTrees = fsTrees;
+    return dependencies;
+  }
+};
+
+/**
+ * Get an FSTree that uses content hashing information to compare files to
+ * see if they have changed.
+ *
+ * @param fsRoot {string}
+ * @param dependencies {Set<string>}
+ * @return {FSHashTree}
+ */
+function getHashTree(fsRoot, dependencies) {
+  /** @type {Array<HashEntry>} */
+  let entries = [];
+  for (let dependency of dependencies) {
+    let fullPath = path.join(fsRoot, dependency);
+    try {
+      // it would be good if we could cache this and share it with
+      // the read that accompanies `processString()` (if any).
+      let contents = fs.readFileSync(fullPath, "utf8");
```

Comment:
The `md5sum` code looks like it supports buffers directly, should we bother encoding this as `utf8` or leaving it as a buffer, so the crypto library can operate on it without encoding.